### PR TITLE
Adding TensorFlow model support extension to Siddhi

### PR DESCRIPTION
**Pull request by niruhan**
## Purpose
Support import of TensorFlow saved models

## Goals
to get inferences from pre-built TensorFlow models using Siddhi

## User stories
https://docs.google.com/a/wso2.com/document/d/1aISBIpG5hzCXzeOh_vK3DSG_B63HT40B4xTZWCFREOM/edit?usp=sharing

## Automation tests
 - Unit tests 
   > Code coverage: 72%

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: component/docs/api/1.0.9-SNAPSHOT.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# API Docs - v1.0.9-SNAPSHOT
```
Comment (pcnfernando) [inline component/docs/api/1.0.9-SNAPSHOT.md]: The version should be v1.0.0. please check

File: component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtension.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+                            .getSignatureDefOrThrow(defaultServingSignatureDefKey);
+        } catch (InvalidProtocolBufferException e) {
+            throw new SiddhiAppCreationException("Error while reading signature def." + e.getMessage(), e);
+        }
+
+        noOfInputs = signatureDef.getInputsCount();
+        noOfOutputs = signatureDef.getOutputsCount();
+
+        //Checking if the specified number of inputs are given
+        final int noOfQueryParams = 1 + 2 * noOfInputs + noOfOutputs;
```
Comment (pcnfernando) [inline component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtension.java:210]: Shall we replace (1+2) this with minConstants

File: docs/api/1.0.7.md [hunk 3]
```diff
@@ -0,0 +1,372 @@
+# API Docs - v1.0.7
```
Comment (pcnfernando) [inline docs/api/1.0.7.md]: Please remove all the invalid docs

Comment (pcnfernando) [reply docs/api/1.0.7.md]: Remove the docs directory

**Review by pcnfernando — CHANGES_REQUESTED** [review]

File: component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtension.java (hunk 2 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtension.java:210]: that is 1 + (2 x noOfInputs)

Comment (pcnfernando) [reply component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtension.java:210]: Shall we add a comment describing what 1 and 2 are

Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtension.java:210]: added

File: component/src/test/resources/log4j.properties [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#
+# Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+#
```
Comment (mohanvive) [inline component/src/test/resources/log4j.properties]: Fix the copyright year.

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+        });
+
+        siddhiAppRuntime.start();
+        InputHandler inputHandler = siddhiAppRuntime.getInputHandler("InputStream");
+        try {
+            inputHandler.send(new Object[]{"float:[1, -2]"});
+            inputHandler.send(new Object[]{"float:[1, 2]"});
+            inputHandler.send(new Object[]{"float:[5, 2]"});
+        } catch (Exception e) {
+            logger.error(e.getCause().getMessage());
```
Comment (mohanvive) [inline component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: Log the complete exception trace and give context details as the error message.

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+                }
+            }
+        });
+
+        siddhiAppRuntime.start();
+        InputHandler inputHandler = siddhiAppRuntime.getInputHandler("InputStream");
+        try {
+            inputHandler.send(new Object[]{"double :[1,-2]"});
+        } catch (Exception e) {
+            logger.error(e.getCause().getMessage());
```
Comment (mohanvive) [inline component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: Log complete exception stack trace.

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+
+            String keepProbString = "float:[";
+
+            for (float num : keepProbArray) {
+                keepProbString  = keepProbString + num + ",";
+            }
+
+            keepProbString = keepProbString.substring(0, keepProbString.lastIndexOf(",")) + "]";
+
+
```
Comment (mohanvive) [inline component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: remove extra line breaks.

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+            String imageAsString = "float:[";
+
+            for (float num : imgAsFloatArray) {
+                imageAsString  = imageAsString + num + ",";
+            }
+
+            imageAsString = imageAsString.substring(0, imageAsString.lastIndexOf(",")) + "]";
+
+
```
Comment (mohanvive) [inline component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: Remove extra line spaces. Check other places as well.

File: component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/util/CoreUtils.java [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+        arrayAsString = arrayAsString.replaceAll("]", " ");
+        arrayAsString = arrayAsString.replaceAll(",", " ");
+        arrayAsString = arrayAsString.replaceAll("\\[", " ");
+        String[] stringArray = arrayAsString.split(" ");
+
+        List tempList = new LinkedList();
+
+        for (String string : stringArray) {
+            if (string != null && string.length() > 0) {
```
Comment (mohanvive) [inline component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/util/CoreUtils.java]: use isEmpty method rather using length to check whether a string is empty or not.

Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/util/CoreUtils.java]: done

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java (hunk 5 above)
Comment (niruhan) [reply component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: done

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java (hunk 8 above)
Comment (niruhan) [reply component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: done

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java (hunk 7 above)
Comment (niruhan) [reply component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: done

File: component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java (hunk 6 above)
Comment (niruhan) [reply component/src/test/java/org/wso2/extension/siddhi/execution/tensorflow/TensorFlowExtensionTest.java]: done

File: component/src/test/resources/log4j.properties (hunk 4 above)
Comment (niruhan) [reply component/src/test/resources/log4j.properties]: done

**Review by mohanvive — APPROVED** [review]

**Comment by mohanvive** [comment]
@pcnfernando  Please validate the changes and approve.

**Review by pcnfernando — APPROVED** [review]

File: component/pom.xml [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Private-Package>
+                            com.google.protobuf.*,
+                            org.tensorflow.*
+                        </Private-Package>
+                        <Export-Package>
+                            org.wso2.extension.siddhi.execution.tensorflow.*
+                        </Export-Package>
+                        <Import-Package>
+                            *;resolution:=optional
```
Comment (ramindu90) [inline component/pom.xml:107]: @niruhan shall we add the exact import packages

File: component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/util/CoreUtils.java [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        }
+        return inputVariableExpressionExecutors;
+    }
+
+    public static List<Attribute> getReturnAttributeList(SignatureDef signatureDef, int noOfOutputs,
+                                                         SavedModelBundle tensorFlowSavedModel,
+                                                         String[] outputNamesArray) {
+        List<Attribute> attributeList = new ArrayList<>(noOfOutputs);
+        for (int i = 0; i < noOfOutputs; i++) {
+
```
Comment (ramindu90) [inline component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/util/CoreUtils.java:72]: @niruhan please remove the unnecessary spaces. check the other files as well.

File: component/pom.xml (hunk 10 above)
Comment (niruhan) [reply component/pom.xml:107]: Fixed in https://github.com/siddhi-io/siddhi-execution-tensorflow/pull/11

File: component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/util/CoreUtils.java (hunk 11 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/execution/tensorflow/util/CoreUtils.java:72]: Fixed in https://github.com/siddhi-io/siddhi-execution-tensorflow/pull/20

**Pull request merged**